File: 2023/cpp/1.cpp
#include <iostream>
#include <string>
#include <fstream>

using namespace std;

int calibration_value(string doc) {
    const string digits = "123456789";
    return stoi(string() + doc[doc.find_first_of(digits)] + doc[doc.find_last_of(digits)]);
}

int true_calibration_value(string doc) {
    // track the minimum index in string and location
    const string numbers[9] = {"one", "two", "three", "four", "five", "six", "seven", "eight", "nine"};
    int min_first = 100000;
    int min_loc = -1;
    int max_last = -10;
    int max_loc = -1;

    // search for each string number in the document
    for (int i = 0; i < 9; i++) {
        int forward_search = doc.find(numbers[i]);
        int backward_search = doc.rfind(numbers[i]);

        // if the number is found then track its index and value
        if (forward_search != -1 && forward_search < min_first) {
            min_first = forward_search;
            min_loc = i;
        }
        if (backward_search != -1 && backward_search > max_last) {
            max_last = backward_search;
            max_loc = i;
        }
    }

    // same as part one search
    const string digits = "123456789";
    int first_digit = doc.find_first_of(digits);
    int last_digit = doc.find_last_of(digits);

    // decide whether to use the digits or string version of the numbers
    string first_num, last_num;
    if (first_digit < min_first && first_digit != -1) {
        first_num = string() + doc[first_digit];
    } else {
        first_num = to_string(min_loc + 1);
    }
    if (last_digit > max_last && last_digit != -1) {
        last_num = string() + doc[last_digit];
    } else {
        last_num = to_string(max_loc + 1);
    }

    // stick em together and sit back in fear at the difficulty of day 1
    return stoi(string() + first_num + last_num);
}

int main(void) {
    string line;
    int total = 0;
    int part_two_total = 0;
    ifstream file("../inputs/1.txt");
    if (file.is_open()) {
        while (getline(file, line))
        {
            total += calibration_value(line);
            part_two_total += true_calibration_value(line);
        }
        file.close();
        cout << "PART ONE: " << total << endl;
        cout << "PART TWO: " << part_two_total << endl;
    } else {
        cout << "Unable to open file"; 
    }
}

File: 2023/julia/1.jl
const digits = [Char('0' + i) for i in 1:9]
const digit_strs = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"]
const replacements = [(digit_strs[i], digit_strs[i][1] * digits[i] * digit_strs[i][length(digit_strs[i])])
                      for i in eachindex(digit_strs)]

function calibration_value(doc::String)
    cdoc = collect(doc)
    first_loc, first_val = 1000000, ""
    last_loc, last_val = -1, ""

    for digit in digits
        mask = cdoc .== digit

        # find earliest occurence of digit in string, if better than current first, update
        earliest_loc = findfirst(mask)
        if earliest_loc !== nothing && earliest_loc < first_loc
            first_loc = earliest_loc
            first_val = digit
        end;

        # find latest occurence of digit in string, if better than current last, update
        latest_loc = findlast(mask)
        if latest_loc !== nothing && latest_loc > last_loc
            last_loc = latest_loc
            last_val = digit
        end;
    end;
    # concatenate and convert to integers
    return parse(Int, first_val * last_val)
end

function main()
    part_one, part_two = 0, 0
    open("../inputs/1.txt") do input
        for line in eachline(input)
            # perform simple part one version
            part_one += calibration_value(line)

            # then replace any string versions of numbers
            # replace every number => its digit, BUT retain first/last characters in case of ellision
            for (before, after) in replacements
                line = replace(line, before => after)
            end;

            # repeat with the new string
            part_two += calibration_value(line)
        end
    end
    @show part_one
    @show part_two
end

main()